File: 515-largest-values/solution.go
/*
BFS看js 自己寫的
DFS看Go 官解
*/
package main

import "gopractice/datastructure/Tree"

type TreeNode = Tree.TreeNode

func largestValues(root *TreeNode) []int {
	var ans []int
	if root == nil {
		return ans
	}
	var dfs func(*TreeNode, int) int
	dfs = func(root *TreeNode, depth int) int {
		if root == nil {
			return 0
		}
		if depth == len(ans) { //檢查是不是第一次到這層
			ans = append(ans, root.Val.(int)) //是就直接塞值
		} else {
			levelMaxNum := max(root.Val.(int), ans[depth]) //不是的話, 就跟之前塞的值比大小
			ans[depth] = levelMaxNum
		}
		dfs(root.Left, depth+1)
		dfs(root.Right, depth+1)
		return 0
	}
	dfs(root, 0) //這題, 第一層是0, 因為 if depth == len(ans) 比較順暢
	return ans
}

/*
17 ~ 22行自己想不到, 看官解才寫出來
https://leetcode.cn/problems/find-largest-value-in-each-tree-row/solutions/1619294/zai-mei-ge-shu-xing-zhong-zhao-zui-da-zh-6xbs/
*/

File: 515-largest-values/solution.js
/*
BFS看js 自己寫的
DFS看Go 官解
 */
const TreeNode = require('../datastructure/Tree/TreeNode');
const buildLevelOrderTree_v2 = require('../datastructure/Tree/TreeDebugger_v2')

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var largestValues = function(root) {
    let ans = []
    if (root === null)
        return ans
    let queue = [root]
    while (queue.length > 0){
        let level = []
        let levelSize = queue.length
        for (let i = 0; i < levelSize; i++) {
            let curNode = queue.shift()
            level.push(curNode.val)
            if (curNode.left != null)
                queue.push(curNode.left)
            if (curNode.right != null)
                queue.push(curNode.right)
        }
        let levelMaxNum = Math.max(...level)
        ans.push(levelMaxNum)
    }
    return ans
};
if (require.main === module){
    const levelOrder = [1,3,2,5,3,null,9]
    const root = buildLevelOrderTree_v2(levelOrder)
    console.log("Ans = ", largestValues(root))
}